Python Runtime - Polishing

bronsa opened the pull request.

bronsa committed:
dead code

bronsa committed:
chore: make sure permission issues don't impact file deletion

bronsa committed:
run python transform within virtual thread, conform to async api

Closes QUE-2302

bronsa committed:
fix

bronsa committed:
chore: lift utility function

bronsa committed:
refactor: use common function for running python-transform

bronsa committed:
style

bronsa committed:
explicitely pass cancel-chan

qnkhuat committed:
fix permissions when running transform as background job

bronsa committed:
Merge remote-tracking branch 'origin/master' into proj-python/s01e02

bronsa committed:
chore: pipe cancel-chan down to python-runner

bronsa committed:
feat: interrupt request on cancel

bronsa committed:
feat: cancel in python server

bronsa committed:
fix tests

bronsa committed:
pass run-id

bronsa committed:
fix cancel endpoint

bronsa committed:
add cancel button

bronsa committed:
fix running-run for transform-id

wotbrew committed:
Add docker-compose.yml with localstack

Makefile/tests now uses docker-compose

bronsa committed:
fix transform-run-cancelation

bronsa committed:
immediately cancel run

crisptrutski committed:
Hacky as all hell - input files use s3

bronsa committed:
cancel working e2e

crisptrutski committed:
Vile vibe verbiage - but it works (locally)

crisptrutski committed:
Makefile creates bucket

bronsa committed:
leave async cancel for sql

crisptrutski committed:
Linter stuff

bronsa committed:
Merge remote-tracking branch 'origin/proj-python/s01e02' into proj-python/s01e02

crisptrutski committed:
s3 config setting

crisptrutski committed:
writing to a tempfile first is a feature - we can close the dwh query faster

bronsa committed:
Merge remote-tracking branch 'origin/proj-python/s01e02' into proj-python/s01e02

crisptrutski committed:
Minor code and object cleanup

bronsa committed:
fix /test error reporting

bronsa committed:
better endpoint name

crisptrutski committed:
rough metadata transfer (just return raw dtype for now)

crisptrutski committed:
get rid of vestigial working_dir

bronsa committed:
add cancel button to test python endpoint

bronsa committed:
add cancel python test endpoing

crisptrutski committed:
Fix for find-replace breaking signed url

bronsa committed:
working button

crisptrutski committed:
Fix for 404s

crisptrutski committed:
silly move, from demo #p

wotbrew committed:
Tests include boto3 now, so add to venv

wotbrew committed:
Logs endpoint vibes

Needs to be destroyed at some future point

bronsa committed:
chore: docstrings in right position

wotbrew committed:
Logs are viewable inline during execution of python transforms

bronsa committed:
remove old setting

bronsa committed:
feat: settings page with s3 configs

bronsa committed:
feat: split settings

bronsa committed:
use new settings

bronsa committed:
remove placeholders

crisptrutski committed:
Auto-update Python transform function signature when table aliases change

- Automatically updates def transform(...) signature when aliases are modified
- Updates docstring Args section with aligned descriptions and table names
- Adds reset button to restore default slugified aliases
- Validates alias input to only allow valid Python identifiers
- Handles multi-line signatures for many parameters (>80 chars)
- Shows visual distinction between custom and auto-generated aliases
- Fixes code editor layout to prevent overlap with Run button

bronsa committed:
eh, rename to appease the munging

crisptrutski committed:
Some minor fixes around docstring rewriting

bronsa committed:
wip: add transform metrics

bronsa committed:
feat: add instrumentation util namespace

bronsa committed:
chore: allow analytics for transforms

bronsa committed:
instrument the job

bronsa committed:
instrument python execution

crisptrutski committed:
more debugging stuff

bronsa committed:
simplify

bronsa committed:
just use run-id

bronsa committed:
snuck back in

crisptrutski committed:
Use system credentials for s3

crisptrutski committed:
Remove sharp hidden failures edges around bucket creation

crisptrutski committed:
lock down script permissions a bit more

crisptrutski committed:
Various fixes for staging and general life

- Measure time properly
- Make endpoint config optional, so it's skipped in staging
- Make auth key config option, so it's skipped in staging
- Skip dev-only settings defaults in staging
- Add prefix, so staging permissions work
- Use short names for Java classes (Cursive lint)
- Remove trivial comments, answer some others
- Fix test-id that broke malli

crisptrutski committed:
Don't need this negative comment

crisptrutski committed:
Fix build - oops

crisptrutski committed:
fix kondo

crisptrutski committed:
fix unit test

crisptrutski committed:
trouble with kms - support only region without custom endpoint

crisptrutski committed:
solve more fe lints

crisptrutski committed:
fix type errors

crisptrutski committed:
more type check fixes - including some smelly optional override stuff

qnkhuat committed:
fix scheduled python run

bronsa committed:
cleanup, remove csv metrics (as it's going to change), generic with-timing

bronsa committed:
fix import typo

bronsa committed:
fix args

bronsa committed:
add logging

bronsa committed:
remove whitespaces

qnkhuat committed:
switch s3 sdk to v2

bronsa committed:
wip: move away from csv upload

bronsa committed:
relax schema

bronsa committed:
chore: fix tests

qnkhuat committed:
fix build error s3 v2

bronsa committed:
creation tests

bronsa committed:
with double dispatch

bronsa committed:
todo

bronsa committed:
move to common ns

bronsa committed:
note

qnkhuat committed:
fix delete transform + table

imrkd committed:
adding sts dependency

crisptrutski committed:
Add default creds provider to pick up STS

qnkhuat committed:
remove hashp

bronsa committed:
drop table instead of truncating

bronsa committed:
fix: don't remove indentation before """

wotbrew committed:
Increase max processes

For me 50 was exceeded today and I could not start transforms. I am
guessing these are starter numbers and we'll need to figure out values
anyway

wotbrew committed:
Wrap stdin/stderr with timestamp prefixes for each line

bronsa committed:
chore: allow empty schemas

crisptrutski committed:
Critical fix: docstring indentation

bronsa committed:
chore: add error

wotbrew committed:
Use timestamps to interleave python stdin/err output in msg

bronsa committed:
type hints

bronsa committed:
fix hints

bronsa committed:
switch to delete-target-table

bronsa committed:
simpler api

bronsa committed:
hack

crisptrutski committed:
Replace python-runner folder with git submodule

bronsa committed:
an in/out type test

bronsa committed:
note

bronsa committed:
fix: timezone aware datetime

bronsa committed:
wip: pass upload_type to manifest

bronsa committed:
automatically parse datetime fields

bronsa committed:
don't go through qp for create-table/drop-table

wotbrew committed:
Read new events jsonl file and use that for the message

bronsa committed:
unused

qnkhuat committed:
CI for Python Runner (#63113)

wotbrew committed:
Remove stdout/err from clj side (single :events output is used now)

crisptrutski committed:
Date times can roundtrip losslessly and custom metadata is supported

wotbrew committed:
Pass run-id to /logs endpoint

bronsa committed:
chore: move insert-from-source into! driver

bronsa committed:
move utils into transforms.util

qnkhuat committed:
add scheduled job e2e tests

bronsa committed:
fix: sort data-rows

bronsa committed:
root-type instead of upload-type

bronsa committed:
add new type->database-type

bronsa committed:
feat: use new methods

bronsa committed:
bump

bronsa committed:
typo

bronsa committed:
simpler

bronsa committed:
fix tests

bronsa committed:
unused

bronsa committed:
feat: per driver impl

bronsa committed:
for snoflake and sqlserver too

bronsa committed:
Merge remote-tracking branch 'origin/master' into proj-python/s01e02

bronsa committed:
fix test

bronsa committed:
more generic types

bronsa committed:
add test

bronsa committed:
fix: cleanup

bronsa committed:
fixes

bronsa committed:
more fixes

crisptrutski committed:
bump python-runner

crisptrutski committed:
delete old test file

bronsa committed:
Merge remote-tracking branch 'origin/master' into proj-python/s01e03

bronsa committed:
fix roundtrip

bronsa committed:
fix bigquery types

bronsa committed:
fix ddl for bigquery

bronsa committed:
add session-schema for snowflake

bronsa committed:
fix for redshift

bronsa committed:
types for clickhouse

bronsa committed:
typo

bronsa committed:
chore: kondo

qnkhuat committed:
Table swapping instead of delete then create

qnkhuat committed:
add tests and polishing

crisptrutski commented on enterprise/backend/src/metabase_enterprise/transforms/execute.clj:
```diff
@@ -228,7 +228,8 @@
 (defn- transfer-file-to-db [driver db {:keys [target] :as transform} metadata temp-file]
   (let [table-name (transforms.util/qualified-table-name driver target)
-        table-schema {:name table-name
+        temp-table-name (str table-name "_temp_" (System/currentTimeMillis))
```
If we're using this method for making names, we should handle conflicts.

crisptrutski replied:
What about using the run id instead?

qnkhuat committed:
proper transform ordering

crisptrutski committed:
Test for ordering across databasees, if we enable python support

crisptrutski committed:
Merge branch 'master' into proj-python/s01e03

bronsa committed:
kondo

bronsa committed:
stub rename-table for bigquery

bronsa committed:
wip: DDL/DML for mongo

bronsa committed:
fix(mongo): guard against lack of cancel-chan

bronsa committed:
only deal with schemas if schema exist

romeovs committed:
Add first-class PythonTransformSource type

romeovs committed:
Lay python editor out in two columns

romeovs committed:
Make python editor resizable

romeovs committed:
Move parseCSV to utils

romeovs committed:
Split off results into separate component

bronsa committed:
chore: bind cancel-chan

bronsa committed:
fix: retrieve meta from appdb

romeovs committed:
Add empty and loading state components

romeovs committed:
Split up ExecutionResult component

romeovs committed:
Move python execution to separate hook

romeovs committed:
Remove color in editor gutter

romeovs committed:
Increase indentUnit for python

bronsa committed:
wip: filter out nested fields

wotbrew committed:
Separate file / s3 lifetimes / writes / reads from http calls (#63294)

- python-runner no longer simulates a single http api that does
everything (incl s3 interaction)
- Separates file lifetimes (shared storage) from execution
- Easier to see how to avoid reading into strings (e.g. for piping into
tmp files, follow up)

Earlier on we had a service that managed much of the I/O, metabase now
coordinates it so it makes little sense pretending there is still
an overall http api.

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/api.clj:
```diff
@@ -1,7 +1,10 @@
+   [clojure.string :as str]
    [metabase-enterprise.transforms.api.transform-job]
    [metabase-enterprise.transforms.api.transform-tag]
+   [metabase-enterprise.transforms.canceling :as trasnforms.canceling]
```
typo

bronsa committed:
feat: allow overriding base_type/db_type

bronsa committed:
kondo

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/api.clj:
```diff
@@ -222,9 +241,16 @@
+    (if (transforms.util/python-transform? transform)
+      (u.jvm/in-virtual-thread*
+       (transforms.execute/execute-python-transform! transform {:start-promise start-promise
+                                                                :message-log   (atom {:pre-python  []
```
Also: Holding all user facing messages in memory is fine for messages we control (few) - but might not be for user code, as we do not know how big the logs will get.

Caveat: The python runner also has a small logs assumption IIRC on the poll side. If we have some truncation/dropping behaviour python side perhaps it is ok to not enforce a dropping buffer / truncation on the JVM side.

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/api.clj:
```diff
@@ -207,11 +223,14 @@
+  (let [transform (api/check-404 (t2/select-one :model/Transform id))
+        run (api/check-404 (transform-run/running-run-for-transform-id id))]
+    (transform-run-cancelation/mark-cancel-started-run! (:id run))
+    (when (transforms.util/python-transform? transform)
```
Add a comment describing why this additional cancellation behaviour exists (optimistic to exit earlier than the 10 minute job on single nodes or if you 'get lucky' with the load balancer).

romeovs committed:
Allow extension to be passed to CodeEditor

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/execute.clj:
```diff
@@ -37,24 +66,38 @@
+      (transform-run/succeed-started-run! run-id)
+      ret)
     (catch Throwable t
       (transform-run/fail-started-run! run-id {:message (.getMessage t)})
```
What should we do with stdout/err output in this case as this overwrites the message field and you might lose information that is essential for debugging the failure.

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/execute.clj:
```diff
@@ -78,24 +121,250 @@
          (deliver start-promise t))
        (throw t)))))
+
+(defn- log! [message-log s]
```
Docstrings/names 

In general need to decide on policy re log truncation (cross cutting across py / jvm)

Recommendation:
- Ensure py code only offers a truncated view of logs on poll. Documented as part of its contract. 
- Add a note alluding to this here so folks do not concern with memory usage

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/execute.clj:
```diff
@@ -78,24 +121,250 @@
+        (log/errorf e "An exception was caught during msg update loop, run-id: %s" run-id)))))
+
+(defn- open-log-thread! ^Closeable [run-id message-log]
+  (let [log-thread-promise (promise)
```
Look at using the cancellable future from in-virtual-thread* and having the same behaviour without thread reference capture

i.e. fatalf is quite important in my view - signal that the process is not responding to interruption and might run forever.

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/execute.clj:
```diff
@@ -78,24 +121,250 @@
+      Closeable
+      (close [_] (cleanup)))))
+
+(defn- debug-info-str [{:keys [exit-code events]}]
```
`message` behaviour needs a general re-think, many such comments here

bronsa committed:
eh, array/dictionary for bigq

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/execute.clj:
```diff
@@ -78,24 +121,250 @@
+        (log/info "Existing table deletected, Create then swap")
+        (try
+          (create-table-and-insert-data! driver (:id db) temp-table-name metadata data-source)
+          ;; TODO: These operations should be within a transaction for atomicity
```
A couple of possible options that come to mind:
- Either comment saying we only care about likelyhood and not any kind of guarantee you will never see the table 'not exist', hence no atomicity guarantee
- Make it driver specific (and open to additional indirection approaches such as using a view)

My 2pence: do not make any such atomicity guarantee. I suspect it might be very hard to achieve across all targets

wotbrew replied:
Even if there is no guarantee a driver method to permit usage of a transaction for cases where it is possible is good in my view, as if you fail after the drop (never rename) its going to be interesting.

Begs the question what the answer is to such situation where we cannot get the atomicity

crisptrutski replied:
yeah i think there should only be some basic guarantees:

1. if there was an output table before, there will be one afterwards
2. the time where the output table ceases to exist will be as vanishingly small as we can make it for the driver. for mysql and postgres thankfully this is "no time at all", but i doubt we can do this for bigquery or mongo.

clean up of temporary junk (either old or new tables) can be best effort, in the case where DDL cannot run in a transaction, or would add unreasonable overhead.

romeovs committed:
Add pandas completions

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/execute.clj:
```diff
@@ -78,24 +121,250 @@
+(defn test-python-transform!
+  "Execute a transform in test mode (does not write result into a table)."
+  [code tables->id run-id cancel-chan]
+  (with-open [shared-storage-ref (python-runner/open-s3-shared-storage! tables->id)]
```
.close is all well and good but the node might shutdown or die (virt/daemon thread!)

Deal with garbage / orphaned objects in storage

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/python_runner.clj:
```diff
@@ -0,0 +1,423 @@
+              (.newLine writer)))
+          reducible-rows)
+
+    (doto writer
```
with-open (.close) should flush anyway

romeovs committed:
Swap buttons

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/python_runner.clj:
```diff
@@ -0,0 +1,423 @@
+   (try
+     (let [^GetObjectRequest request (build-get-object-request bucket-name key)
+           response                  (.getObject s3-client request)]
+       (slurp response))
```
they also might not all be utf-8 and slurp do anything useful

wotbrew replied:
files might be any old binary right? Just thinking this function in particular (assuming all objects represent strings, perhaps to be subsequentally parsed as text)

wotbrew replied:
I've already made the point elsewhere about reading the objects eagerly - I do not expect this function to survive as is

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/python_runner.clj:
```diff
@@ -0,0 +1,423 @@
+                   (BufferedWriter.))]
+
+    (run! (fn [row]
+            (let [row-map (zipmap col-names row)]
```
do we exit early here if > too many rows visited / too long

bronsa committed:
fixes for bigq insertion

wotbrew commented on src/metabase/upload/core.clj:
```diff
@@ -7,9 +7,11 @@
 
 (p/import-vars
  [metabase.upload.impl
+  create-from-csv-and-sync!
```
can remove these new exports I guess now we do not use the csv call

bronsa committed:
fixes for bigq rename-table

wotbrew commented on src/metabase/driver.clj:
```diff
@@ -1294,6 +1304,31 @@
+(defmulti insert-from-source!
+  "Inserts data from a data source into an existing table.
+   This abstracts away the conversion and insertion process, allowing drivers
+   to optimize based on the data source type. Returns number of rows inserted. "
```
Extra docs / definition on 
- What are the requirements of implementers in terms of interpreting data sources (data source types they must implement)
- Is data being written expected to be visible to readers? (Or is a particular transaction + isolation level semantics being prescribed)
- If the operation fails is the implementation allowed to leave data in the table? (Same ACID analysis as above answers both I expect)
- Is there any limit to the size of the data source?
- Is there any requirement for implementations to timeout?
- Is the call expected to block and return only once finished with the data source (so callers must guarantee it exists only during the call)

wotbrew replied:
Totally agree - I would state it explicitly though 'Callers are free to X, are free to Y' etc

wotbrew commented on src/metabase/driver.clj:
```diff
@@ -1294,6 +1304,31 @@
+(defmethod insert-from-source! [::driver :rows]
+  [driver db-id table-name column-names {:keys [data]}]
+  (when (seq data)
+    (insert-into! driver db-id table-name column-names data))
```
Likely answers the transactional / failure mode question above

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/python_runner.clj:
```diff
@@ -0,0 +1,423 @@
+        payload                  {:code                code
+                                  :timeout             30
+                                  :request_id          run-id
+                                  :output_url          (:url output)
```
For all storage locations we should specify explicitly the intended version and format so that the python service is able to negotiate the formatting depending on the version of metabase that is running. One option might be to version the http request as a whole with the metabase version or some other surrogate rather than having a multitude of independent version numbers.

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/execute.clj:
```diff
@@ -78,24 +121,250 @@
+          {:keys [exit_code]} body
+          ;; TODO temporary to keep more code stable while refactoring
+          ;; no need to materialize these early (i.e output we can stream directly into a tmp file or db if small)
+          {:keys [output output-manifest events]} (python-runner/read-output-objects @shared-storage-ref)]
```
Address notes - discussed in other comments

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/execute.clj:
```diff
@@ -78,24 +121,250 @@
+              {:db db
+               :transform-type (keyword (:type target))
+               :conn-spec (driver/connection-spec driver db)
+               ;; :query (transforms.util/compile-source source)
```
delete commented out code

romeovs committed:
Use DataGrid for displaying results

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/api.clj:
```diff
@@ -234,6 +260,53 @@
+            [:tables [:map-of :string :int]]]]
+  (log/info "test python code execution")
+  (api/check-superuser)
+  (let [run-id python-test-run-id
```
> re: nodes, i think we're assuming that either the worker is a singleton or we have a sticky customer session. don't think concurrent transforms for a given customer are in scope for v1, but haven't got that in writing.

I meant metabase nodes if that helps?

wotbrew commented on src/metabase/driver.clj:
```diff
@@ -1294,6 +1304,31 @@
+
+(defmethod insert-from-source! [::driver :csv-file]
+  [driver db-id table-name column-names {:keys [file]}]
+  (let [csv-rows (csv/read-csv (io/reader file))
```
On a more careful read, this already opens the handle and hands a lazy seq off to insert-from-source!. The lifetime of the file handle is not explicit though, presumably it is closed once the seq reaches an EOF as part of the lazy seq offered by `read-csv`. But its not clear.
- If we are going to use the lazy seq approach we should use `with-open` to limit surprise.
- On error right now I'm not sure the file will be closed (maybe once the finalizer of the reader runs)

wotbrew commented on src/metabase/driver.clj:
```diff
@@ -1294,6 +1304,31 @@
+  (let [csv-rows (csv/read-csv (io/reader file))
+        header-rows (first csv-rows)
+        data-rows (map (fn [rows]
+                         (let [m (zipmap header-rows rows)]
```
ditto timeout

qnkhuat committed:
Start python runner with driver tests (#63335)

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/python_runner.clj:
```diff
@@ -0,0 +1,423 @@
+  [run-id]
+  (let [server-url (transforms.settings/python-execution-server-url)]
+    (http/get (str server-url "/logs")
+              {:content-type     :json
```
(though log thread I suppose is in the background, and can be interrupted anyway) - might not be a problem to allow this to block for as long as it wants

wotbrew committed:
Remove some commented out / dead code

crisptrutski committed:
Pin to a fixed python-runner, with slightly changed logs

bronsa committed:
fix support for clickhouse

bronsa committed:
feat: json

crisptrutski commented on enterprise/backend/src/metabase_enterprise/transforms/execute.clj:
```diff
@@ -78,24 +121,250 @@
+(defn execute-python-transform!
+  "Execute a Python transform by calling the python runner.
+
+  This is executing synchronously, but supports being kicked off in the background
```
I _think_ this comment is referring to the fact that it could be called from both the API or Quartz.

bronsa committed:
fix for clickhouse driver

bronsa committed:
assoc table-schema to data-source

wotbrew committed:
Add a few comments regarding optimistic nature of s3 deletes

bronsa committed:
feat: decode json strings

crisptrutski committed:
Bump to latest submodule

bronsa committed:
fix: schema-exists? check

bronsa committed:
chore: Dictionary as root type

bronsa committed:
bump

bronsa committed:
more parsing for bigq

bronsa committed:
typo

bronsa committed:
format datetimes for bigq

bronsa committed:
type conversion for mongo

bronsa committed:
missing require

bronsa committed:
chore

bronsa committed:
sqlserver rename-table

bronsa committed:
fix(ci): make run

wotbrew committed:
Interop with future task rather than doing old-school thread stuff on log process shutdown

wotbrew committed:
Remove unused import

bronsa committed:
Merge remote-tracking branch 'origin/master' into proj-python/s01e03

bronsa committed:
typo

bronsa committed:
bump

bronsa committed:
typo

bronsa committed:
feat: better insert-from-source abstraction

bronsa committed:
reintroduce fix for floats on bigquery

Closes QUE-2444

bronsa committed:
feat: move driver-specific hacks during preprocessing

crisptrutski committed:
Upgrade to latest python-runner

bronsa committed:
docstrings

bronsa committed:
fix

qnkhuat committed:
Python library APIs (#63378)


Co-authored-by: Romeo Van Snick <romeo@romeovansnick.be>

qnkhuat committed:
Atomically swapping tables for python transform (#63292)

bronsa committed:
Python jsonl (#63437)

* feat: jsonl

* bump

romeovs committed:
Use normal title case

romeovs committed:
Add syntax highlighting to Python script on Transform page

romeovs committed:
Handle json-line output

romeovs committed:
Remove unused props

romeovs committed:
Rename PythonEditor to PythonQueryEditor

romeovs committed:
Move script rewriter to utils

romeovs committed:
Flatten Python transform page

romeovs committed:
Rename shared to common

romeovs committed:
Remove dead code

romeovs committed:
Reduce specialization in Transform editor

romeovs committed:
Remove unused props

romeovs committed:
Move PythonQueryEditor to top-level components

romeovs committed:
Nest Python editor components in PythonTransformEditor

romeovs committed:
Restructure Python transform editor

romeovs committed:
Split up component more

romeovs committed:
Add helpers for python editor

romeovs committed:
Disable default schema

romeovs committed:
Add test ids

romeovs committed:
Fix import path

romeovs committed:
Reuse transform save modal

romeovs committed:
Add e2e test for Python transform cretion

romeovs committed:
Fix types

romeovs committed:
Fix parameter

romeovs committed:
Add e2e tests for python common lib

romeovs committed:
Move python transform api to general transforms folder

romeovs committed:
Move type to metabase-api/types

romeovs committed:
Disable debugger for python transforms for now, allow to open it with query param

romeovs committed:
Disable debugger hotkey for python transforms
